### PR TITLE
Add means to by-pass readiness checks

### LINKED ISSUE #7224: Means to by-pass readiness checks in reconciliation flow
**What would you like to be added**:
The reconciliation flow may get stuck if certain readiness probes fail, e.g. for ETCD or the API servers. This is especially a problem during zone outages and when end users may want to "push through" certain counter-measures like: adding a new zone, switch the machine type or enlarging a worker pool, reconfigure the cluster autoscaler. This is all not possible and will be blocked by the readiness probes that won't succeed if not all pods are ready, which the ones in the broken zone never will.

We discussed out-of-band various options, e.g.:
- Wait for the pods in just two zones to become ready, but w/o knowledge of a zone outage and if we do that always, this won't work as then follow-up steps would possibly hit old API servers and if that's about a minor version upgrade that may fail silently like in the case with cluster autoscaler and v1.25
- Wait for ETCD quorum, but w/o knowledge of a zone outage and if we do that always, this will have catastrophic consequences if we do credentials rotation step 2 as then the third pod gets locked out
- Refactoring the huge reconcile flow into different independent reconcilers, but that's a very complex task and bears the risk to easily introduce bugs as what is relatively "simple" today, would become a daunting task tomorrow, if every reconciler must then check it's prerequisites in detail (instead of relying on the dependent steps to have executed successfully, which the flow guarantees today)

So, the suggestion was to first start simple/small and implement an annotation, that if set, suppresses all readiness checks in the reconcile flow. This way, whoever needs to "push through" certain spec changes, also during a zone outage, can always do so.

**Why is this needed**:
To unblock cluster admins that need to "push through" certain spec changes (examples see above), also and especially during a zone outage where we already know today, that our readiness checks would block those.

### PR DESCRIPTION
**Pull request by timuthy**
**How to categorize this PR?**

/area control-plane
/area high-availability
/kind enhancement

**What this PR does / why we need it**:
This PR adds the `shoot.gardener.cloud/skip-readiness` annotation which, if set, skips most of the readiness checks in the shoot reconcile flow. Please see #7224 for more information.

**Which issue(s) this PR fixes**:
Fixes #7224

**Special notes for your reviewer**:
Some steps are still not skipped:

- `waitUntilKubeAPIServerServiceIsReady` -> Won't help to skip this step since a working KAS service is the foundation for most actions
- `waitUntilControlPlaneExposureDeleted` -> Only relevant if changed from non-SNI to SNI and will (probably) be gone soon
- `waitUntilOperatingSystemConfigReady` -> Will lead to a nil pointer exception since this step extracts data from the `OSC`'s status that is consumed by `DeployManagedResourceForCloudConfigExecutor`.

**Release note**:

```feature operator
A new shoot annotation `shoot.gardener.cloud/skip-readiness` has been added. Gardener skips most readiness checks in the shoot reconciliation flow when shoots have this annotation. It is meant to push through shoot spec changes in case of critical situations, e.g. an availability zone outage, in which various steps can never reach readiness. Once successfully reconciled, the annotation is automatically removed again. Using this annotation however, must be evaluated on a case-by-case basis since it can severely affect the availability of shoot control planes.
```

### CODE REVIEW

File: docs/usage/shoot_high_availability.md [hunk 1]
```diff
@@ -49,4 +49,16 @@ If you already have a shoot cluster with non-HA control plane then following upg
 
 **Disallowed Transitions**
 
-If you have already set-up a HA shoot control plane with `node` failure tolerance then an upgrade to `zone` failure tolerance is currently not supported, mainly because already existing volumes are bound to the zone they were created in.
\ No newline at end of file
+If you have already set-up a HA shoot control plane with `node` failure tolerance then an upgrade to `zone` failure tolerance is currently not supported, mainly because already existing volumes are bound to the zone they were created in.
```
Comment (rfranzke) [inline docs/usage/shoot_high_availability.md]:
```suggestion
If you have already set-up an HA shoot control plane with `node` failure tolerance, then an upgrade to `zone` failure tolerance is currently not supported, mainly because already existing volumes are bound to the zone they were created in originally.
```

File: docs/usage/shoot_high_availability.md [hunk 2]
```diff
@@ -49,4 +49,16 @@ If you already have a shoot cluster with non-HA control plane then following upg
...
+
+## Zone Outage Situation
+
+An availability zone outage might lead to the requirement to change your cluster setup temporarily and on short notice, in order to compensate failures and shortages resulting from the outage.
+For instance, if the shoot cluster has worker nodes across three zones where one zone goes down, the computing power from these nodes is also gone during that time.
+Changing the worker pool (`shoot.spec.provider.workers[]`) and infrastructure (`shoot.spec.provider.infrastructureConfig`) configuration can eliminate this disbalance, having enough machines in healthy availability zones that can cope with the requests of your applications.
+
+Gardener relies on a sophisticated reconciliation flow with several dependencies for which various flow steps wait for the _readiness_ of prior ones.
+During a zone outage this can block the entire flow, e.g. because all three `etcd` replicas can never be ready when a zone is down, and required changes mentioned above can never be accomplished.
```
Comment (rfranzke) [inline docs/usage/shoot_high_availability.md]:
```suggestion
During a zone outage, this can block the entire flow, e.g. because all three `etcd` replicas can never be ready when a zone is down, and required changes mentioned above can never be accomplished.
```

File: docs/usage/shoot_high_availability.md [hunk 3]
```diff
@@ -49,4 +49,16 @@ If you already have a shoot cluster with non-HA control plane then following upg
...
+For this, a special one-off annotation `shoot.gardener.cloud/skip-readiness` helps to skip any readiness checks in the flow.
+
+> The `shoot.gardener.cloud/skip-readiness` annotation serves as a last resort if reconciliation is stuck because of important changes during an AZ outage. Use it with caution, only in exceptional cases and after a case-by-case evaluation with your Gardener landscape administrator. If used with other operations like Kubernetes version upgrades or credential rotation, the annotation may lead to a severe outage of your shoot control plane.
```
Comment (rfranzke) [inline docs/usage/shoot_high_availability.md]:
```suggestion
> The `shoot.gardener.cloud/skip-readiness` annotation serves as a last resort if reconciliation is stuck because of important changes during an AZ outage. Use it with caution, only in exceptional cases and after a case-by-case evaluation with your Gardener landscape administrator. If used together with other operations like Kubernetes version upgrades or credential rotation, the annotation may lead to a severe outage of your shoot control plane. 
```

File: test/e2e/gardener/shoot/create_update_delete.go [hunk 4]
```diff
@@ -108,6 +110,32 @@ var _ = Describe("Shoot Tests", Label("Shoot", "default"), func() {
 			Shoot:             f.Shoot,
 		}, nil, nil)
 
+		By("Add skip readiness annotation")
+		ctx, cancel = context.WithTimeout(parentCtx, 10*time.Minute)
+		defer cancel()
+		Expect(f.ShootFramework.UpdateShoot(ctx, func(shoot *gardencorev1beta1.Shoot) error {
+			metav1.SetMetaDataAnnotation(&shoot.ObjectMeta, "shoot.gardener.cloud/skip-readiness", "")
+			// Use maintain operation to also execute skipped steps in the reconcile flow.
```
Comment (rfranzke) [inline test/e2e/gardener/shoot/create_update_delete.go]:
```suggestion
			// Use maintain operation to also execute tasks in the reconcile flow which are only performed during maintenance.
```

File: test/e2e/gardener/shoot/create_update_delete.go [hunk 5]
```diff
@@ -108,6 +110,32 @@ var _ = Describe("Shoot Tests", Label("Shoot", "default"), func() {
...
+			metav1.SetMetaDataAnnotation(&shoot.ObjectMeta, "gardener.cloud/operation", "maintain")
+			return nil
+		})).To(Succeed())
+
+		By("Wait for operation annotation to be gone. Controller picked up reconciliation request.")
```
Comment (rfranzke) [inline test/e2e/gardener/shoot/create_update_delete.go]:
```suggestion
		By("Wait for operation annotation to be gone (meaning controller picked up reconciliation request)")
```

File: pkg/gardenlet/controller/shoot/shoot/reconciler_reconcile.go [hunk 6]
```diff
@@ -671,8 +678,8 @@ func (r *Reconciler) runReconcileShootFlow(ctx context.Context, o *operation.Ope
 			Dependencies: flow.NewTaskIDs(hibernateControlPlane),
 		})
 		_ = g.Add(flow.Task{
-			Name:         "Waiting until extension resources hibernated after kube-apiserver hibernation are ready",
```
Comment (rfranzke) [inline pkg/gardenlet/controller/shoot/shoot/reconciler_reconcile.go:680]: Why was this removed?

File: pkg/gardenlet/controller/shoot/shoot/reconciler_reconcile.go [hunk 7]
```diff
@@ -757,6 +764,14 @@ func (r *Reconciler) runReconcileShootFlow(ctx context.Context, o *operation.Ope
 		}
 	}
 
+	o.Logger.Info("Removing one-off annotations")
```
Comment (rfranzke) [inline pkg/gardenlet/controller/shoot/shoot/reconciler_reconcile.go]:
```suggestion
	o.Logger.Info("Removing skip-readiness annotation")
```

I don't think we will have others, and if we can still refine this statement. For now, it's clearer to make it more concrete.
However, it would be nice if this is only logged when really done (to not pollute the log/confuse operators).

File: docs/usage/shoot_high_availability.md (hunk 1 above)
Comment (timuthy) [reply docs/usage/shoot_high_availability.md]: I really have no idea why this is shown as a change but I will apply your suggestion 😄

**Review by rfranzke — COMMENTED** [review]
/lgtm
/approve
